sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,538 @@
+```
+// Named array of properties.
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+```
+It contains name of property set and array of properties. Each property is
+represented by the following struct:
+```
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+List of imported symbols is represented as a single property set with name
+`SYCL/imported symbols` recorded in the `Name` field of property set.
+Each property in this set holds name of the particular imported symbol recorded
+in the `Name` field of the property.
+List of exported symbols is represented in the same way, except the
+corresponding set has the name `SYCL/exported symbols`.
+
+### DPC++ runtime changes
+
+DPC++ RT performs *device images collection* task by grouping all device
+images required to execute a kernel based on the list of exports/imports and
+links them together using PI API.
+This native device image is then added to the cache to avoid symbol resolution,
+compilation, and linking for any future attempts to invoke kernels from this
+device image.
+
+#### DPC++ runtime plugin interface (PI) changes
+
+Before creating a program the function `piextDeviceSelectBinary` is used to
+choose the most appropriate device image. It is possible that not all backends
+have possibility to link device images of particular format at run-time. So,
+in presence of dynamic linking the `piextDeviceSelectBinary` function should be
+extended, so it chooses the appropriate device image
+based on additional attributes of device images, such as:
+- List of imports (if present)
+- Device image format
+- Runtime linking support in corresponding backend
+
+Example: the backend doesn't have support of native binaries linking at
+run-time but linking of SPIR-V device images is supported,
+the AOT-compiled device image with required kernel have imports
+information attached which effectively means that this device image needs runtime
+linking, but since native binaries linking is not supported, the image with
+SPIR-V format will be chosen (or an error emitted if there is no other device
+images).
+To link several device images together `piProgramLink` API will be used.
+Depending on concrete plugin implementation and set of device image formats that
+can be linked at run-time, `piProgramLink` API may receive device images in
+different states as inputs (including SPIR-V and native code) with a limitation
+that all inputs should have the same format.
+
+##### Support of runtime linking in backends
+
+- The initial design will support dynamic linking of device code in SPIR-V
```

Comment:
I think it is better say `initial implementation`, because the rest of design doesn't change with updates of a PI plugin implementation.